From 925182953a9541a5cf33ac782327b2cb2bf2a22b Mon Sep 17 00:00:00 2001
From: Dunrong Huang <riegamaths@gmail.com>
Date: Wed, 23 Dec 2015 22:11:10 +0800
Subject: [PATCH 16/23] rgw: add a new error code for invalid key type

From http://docs.ceph.com/docs/master/radosgw/adminops/ when invalid key type
specified, rgw should return InvalidKeyType.

Signed-off-by: Dunrong Huang <riegamaths@gmail.com>
(cherry picked from commit ad1e6489c29bd4fc000d34e1669d90f097b0ad19)
Signed-off-by: Dunrong Huang <riegamaths@gmail.com>
---
 src/rgw/rgw_common.h      | 1 +
 src/rgw/rgw_http_errors.h | 1 +
 src/rgw/rgw_user.cc       | 2 +-
 3 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/rgw/rgw_common.h b/src/rgw/rgw_common.h
index 27c4469..3bba227 100644
--- a/src/rgw/rgw_common.h
+++ b/src/rgw/rgw_common.h
@@ -142,6 +142,7 @@ using ceph::crypto::MD5;
 #define ERR_EMAIL_EXIST          2032
 #define ERR_KEY_EXIST            2033
 #define ERR_INVALID_SECRET_KEY   2034
+#define ERR_INVALID_KEY_TYPE     2035
 #define ERR_USER_SUSPENDED       2100
 #define ERR_INTERNAL_ERROR       2200
 
diff --git a/src/rgw/rgw_http_errors.h b/src/rgw/rgw_http_errors.h
index 8dd9c53..52de232 100644
--- a/src/rgw/rgw_http_errors.h
+++ b/src/rgw/rgw_http_errors.h
@@ -51,6 +51,7 @@ const static struct rgw_http_errors RGW_HTTP_ERRORS[] = {
     { ERR_EMAIL_EXIST, 409, "EmailExists" },
     { ERR_KEY_EXIST, 409, "KeyExists"},
     { ERR_INVALID_SECRET_KEY, 400, "InvalidSecretKey"},
+    { ERR_INVALID_KEY_TYPE, 400, "InvalidKeyType"},
     { ENOTEMPTY, 409, "BucketNotEmpty" },
     { ERR_PRECONDITION_FAILED, 412, "PreconditionFailed" },
     { ERANGE, 416, "InvalidRange" },
diff --git a/src/rgw/rgw_user.cc b/src/rgw/rgw_user.cc
index 01596e4..8f6021f 100644
--- a/src/rgw/rgw_user.cc
+++ b/src/rgw/rgw_user.cc
@@ -904,7 +904,7 @@ int RGWAccessKeyPool::modify_key(RGWUserAdminOpState& op_state, std::string *err
     break;
   default:
     set_err_msg(err_msg, "invalid key type");
-    return -EINVAL;
+    return -ERR_INVALID_KEY_TYPE;
   }
 
   if (!op_state.has_existing_key()) {
-- 
2.8.0.rc3

